Add request-reply use case [16609] (#436)
eProsima/Fast-DDS-docs: docs/fastdds/use_cases/request_reply/request_reply.rst

        cloud "Processing" as Processing
        [Request DataReader] -right-> Processing
        Processing -right-> [Reply DataWriter]
    }

    node "Client application" {
        [Reply DataReader]
        [Request DataWriter]
    }

    [Request DataWriter] -down-> [Request DataReader] : Request Topic
    [Reply DataWriter] --> [Reply DataReader] : Reply Topic

The key for making *Request-Reply* work is relating the Request with the Reply in the client's side.
*Fast DDS* API provides |SampleIdentity-api| to achive this.

A full example can be found in `Fast DDS repository`_.

Getting started
---------------

For *Request-Reply* communication perform the following steps:

1. Define two structures in an IDL file.
   One structure will be used as Request Topic's data type and the other one as Reply Topic's data type.

   .. code-block:: idl

       enum OperationType
       {
           ADDITION,
           SUBSTRACTION,
           MULTIPLICATION,
           DIVISION
       };

       struct RequestType
       {
           OperationType operation;
           long x;
           long y;
       };

       struct ReplyType
       {
           long long z;
       };

> 2. In the client application creates a DataWriter for the request and a DataReader for the reply.

Author: Done in ba11b18e

    .. literalinclude:: ../../../../code/DDSCodeTester.cpp
       :language: c++
       :dedent: 4
       :start-after: //REQUEST_REPLY_EXAMPLE_CLIENT_CREATE_ENTITIES
       :end-before: //!

> 3. In the server application creates a DataWriter for the reply and a DataReader for the request.

Author: Done in ba11b18e

Sending the request and storing the assigned identifier
-------------------------------------------------------

For sending the request, the client application should retrieve and store the internal identifier assigned to the
published sample.
Therefore the sample should be published using the overloaded |DataWriter::write-api| function which second argument is
a reference to a |WriteParams-api| object.
The assigned identifier will be stored in the |WriteParams-api|'s attribute |WriteParams-sample_identity-api|.

.. literalinclude:: ../../../../code/DDSCodeTester.cpp
   :language: c++
   :dedent: 4
   :start-after: //REQUEST_REPLY_EXAMPLE_CLIENT_RETRIEVE_ID
   :end-before: //!

Receiving the request and sending a reply associated with it
------------------------------------------------------------

When the server application receives the request (for example through |DataReaderListener::on_data_available-api|),
it has to retrieve the request's identifier using |SampleInfo::sample_identity-api|.

.. literalinclude:: ../../../../code/DDSCodeTester.cpp
   :language: c++
   :dedent: 4
   :start-after: //REQUEST_REPLY_EXAMPLE_SERVER_GET_ID
   :end-before: //!

> After processing the request, the server should send the reply and it have to be associated with the requestt assigning
> the stored identifier in |WriteParams-related_sample_identity-api|.

Reviewer: ```suggestion
After processing the request, the server should send the reply to the client with the related request attached.
This is done assigning the stored identifier in |WriteParams-related_sample_identity-api|.
```

.. literalinclude:: ../../../../code/DDSCodeTester.cpp
   :language: c++
   :dedent: 4
   :start-after: //REQUEST_REPLY_EXAMPLE_SERVER_SEND_REPLY
   :end-before: //!

Identifying the reply for the client
------------------------------------

When the client application receives a reply (for example through |DataReaderListener::on_data_available-api|),

> the client application should identify which reply is an answer for its request.

Reviewer: ```suggestion
the client application should identify if the reply received is the one expected for its request.
```

Author: Done in ba11b18e

> For this the client application have to compare the stored |SampleIdentity-api| with the incomming

Reviewer: ```suggestion
For this the client application has to compare the stored |SampleIdentity-api| with the incoming
```
